technical-docs/designs/infra/managed-identity.md

```diff
@@ -0,0 +1,251 @@
+Tests for the following modules:
+
+- AKS
+- Key Vault
+- VNet
+- Subnet
+- Gitops
+
+#### Integration Tests
+
+Integration tests will validate resource interoperability upon deployment.
+Pending a successful `terraform apply`, using a go script and terratest go
+library, this design will create an integration test for the respective
+environment template that verifies
+
+- Flux namespace
+- Access to cluster using Pod Identity
+- Access to key using flex-volume
+- 400 response on Voting App
+
+#### Acceptance Test
+
+Acceptance tests are defined in this design as a system affirmation that the
+incoming PR has a successful build in a live staging environment once applied.
+Maintain a live QA environment that successful builds from an incoming PR are
+applied to the state file.
+
+Questions & Limitations:
+
+- With an incoming change to an azure provider module, how will this be applied
+  to an existing terraform deployment. If fail, should we redeploy a new
+  `azure-MI` environment for QA?
+
+#### Reporting
+
+Output a test failure report using out-of-box terratest JUnit compiler to
+capture errors thrown during build.
+
+The whitelisted integration test for `azure-MI` will include:
+
+> `go test -v -run TestIT_Bedrock_AzureMI_Test -timeout 99999s | tee TestIT_Bedrock_AzureMI_Test.log`
+
+> `terratest_log_parser -testlog TestIT_Bedrock_AzureSimple_Test.log -outputdir single_test_output`
+
+The pipeline will publish the XML report as an artifact that is uniquely named
+to AzDO.
+
+```
+ task: PublishPipelineArtifact@1
+        inputs:
+          path: $(modulePath)/test/single_test_output
+          artifact: simple_test_logs
+        condition: always()
+      - task: PublishTestResults@2
+        inputs:
+          testResultsFormat: 'JUnit'
+          testResultsFiles: '**/report.xml'
+          searchFolder: $(modulePath)/test
+        condition: and(eq(variables['Agent.JobStatus'], 'Succeeded'), endsWith(variables['Agent.JobName'], 'Bedrock_Build_Azure_MI'))
+```
+
+## 4. Dependencies
+
+This design for a Managed Identity AKS Testing Harness will leverage the
+following:
+
+- [Bedrock Pre-Reqs: az cli | terraform | golang | fabrikate ](https://github.com/microsoft/bedrock/tree/master/tools/prereqs)
+- [Terratest](https://github.com/gruntwork-io/terratest)
+- [Terraform Compliance](https://github.com/eerkunt/terraform-compliance)
+- [Cobalt Terraform Test Fixtures](https://github.com/microsoft/cobalt/tree/master/test-harness)
+
+## 5. Risks & Mitigations
+
+Risks & Limitations:
+
+- With the deployment of the `azure-common-infra` template for Key Vault, will
+  that also need to be modified for Manage Identity to whitelist AKS to access
+  keyvault?
+- With an incoming change to an azure provider module, how will this be applied
+  to an existing terraform deployment. If fail, should we redeploy a new
+  `azure-MI` environment for QA?
+- How long does it take to deploy MI and Keyvault in a pipeline?
+
```

Comment:
Another risk worth mentioning that flex volume doesnt support MSI enabled clusters today. The only workaround is to create a SP in advance and configure that AKS cluster using that principal, which conflicts with AKS MSI [best practices](https://github.com/Azure/AKS/releases/tag/2020-03-16).

Reply:
I haven't tested but would this address this: https://github.com/Azure/kubernetes-keyvault-flexvol/issues/185

Reply:
maybe? assigning the identity to the vmss will be an automation challenge as we don't have a handle to the vmss in terraform once the cluster is provisioned. I think it's time to look into a CSI migration strategy, which we may want to add to our backlog.